### DISCUSSION TITLE
Use vfs to create mjModel from MJCF's XML already stored in memory?

### DISCUSSION THREAD

**Discussion by traversaro** [google-deepmind/mujoco#24] [Asking for Help] [answered]
First of all, thanks a lot for making MuJoCo available for free and soon open source! 

To test MuJoCo, I tried to set up a proof of concept integration between MuJoCo and [Ignition Gazebo](https://ignitionrobotics.org), see https://github.com/ignitionrobotics/ign-physics/issues/299 for more details. 

While working on that, I had to generate a MJCF file on the file in memory via TinyXML2 . I then wanted to generate a `mjModel` out of this XML file that I had loaded in memory, but I had to write it on a temporary file (see https://github.com/traversaro/ign-physics/blob/b44553c2b2cde6c4ee2cecade2910ef89be8f88e/mujoco/lib/src/World.cc#L325) as `mj_loadXML` function only permits to specify the MJCF's XML via a file name, not via a in-memory string that contains the XML. 

Initially, I tought it would be great to have some way to obtain an `mjModel` via an in-memory MJCF's XML string. Then I checked the docs, and apparently this should be already possible via the [vfs functionality](https://mujoco.readthedocs.io/en/latest/APIreference.html#virtual-file-system): 
> Model and resource files in the VFS can also be constructed programmatically (say using an XML library that writes to memory). 

However, among the vfs function I could not find any that creates a vfs file from a string or buffer present already present in memory. Also the document common usage pattern seem to refer to loading file from the disk in the vfs, not load data that is already in memory in the vfs: 

> The common usage pattern is to first clear it with mj_defaultVFS, then add disk files to it with mj_addFileVFS (which allocates memory buffers and loads the file content in memory), then call mj_loadXML or mj_loadModel, and then clear everything with mj_deleteVFS.

I wonder if I am missing something myself, thanks a lot in advance!

**Comment by yuvaltassa** [comment, marked as answer]
I believe the correct workflow in this case is to create an empty VFS "file" using [`mj_makeEmptyFileVFS`](https://github.com/deepmind/mujoco/blob/main/include/mujoco.h#L110) with the appropriate size, and then copy the contents into the allocated memory space. Does that make sense? 
Note I've never tried this, so would love to hear if it works or not.

**Reply by nimrod-gileadi** [reply]
You should also look at the struct [mjVFS](https://mujoco.readthedocs.io/en/latest/APIreference.html#mjvfs). This would be something along the lines of:
```
mjVFS vfs;
mj_defaultVFS(&vfs);
// TODO: Check return value
mj_makeEmptyFileVFS(&vfs, "filename.xml", xml.size() + 1);
// TODO: Check file_idx is non-negative
int file_idx = mj_findFileVFS(&vfs, "filename.xml");
memcpy(vfs.filedata[file_idx], xml_data, xml.size() + 1);
mjModel *model = mj_loadXML("filename.xml", &vfs, error, error_size);
mj_deleteFileVFS(&vfs, file);
```

**Reply by traversaro** [reply]
Thanks @yuvaltassa and @nimrod-gileadi, that is much more clear! To clarify the (probably obvious) point that I was missing, it was not clear to me that the `mjVFS` struct was itself part of the public API, and not a structure that was only meant to be accessed via `mj_` functions.

**Comment by JaechangKo** [comment]
Hello y'all... 
I think I'm having similar issue like you. 
I have two errors.. and those are,

EntryPointNotFoundException: mj_makeEmptyFileVFS
Mujoco.MjVfs.AddFile (System.String filename, System.String contents) (at D:/Unity/mujoco-main/mujoco-main/unity/Runtime/Tools/MjVfs.cs:50)
Mujoco.MjEngineTool.LoadModelFromString (System.String contents) (at D:/Unity/mujoco-main/mujoco-main/unity/Runtime/Tools/MjEngineTool.cs:64)
Mujoco.MjScene.CompileScene (System.Xml.XmlDocument mjcf, System.Collections.Generic.IEnumerable`1[T] components) (at D:/Unity/mujoco-main/mujoco-main/unity/Runtime/Components/MjScene.cs:167)
Mujoco.MjScene.CreateScene (System.Boolean skipCompile) (at D:/Unity/mujoco-main/mujoco-main/unity/Runtime/Components/MjScene.cs:160)
Mujoco.MjScene.Start () (at D:/Unity/mujoco-main/mujoco-main/unity/Runtime/Components/MjScene.cs:80)

NullReferenceException: Failed to create Mujoco runtime.
Mujoco.MjScene.StepScene () (at D:/Unity/mujoco-main/mujoco-main/unity/Runtime/Components/MjScene.cs:325)
Mujoco.MjScene.FixedUpdate () (at D:/Unity/mujoco-main/mujoco-main/unity/Runtime/Components/MjScene.cs:88)

Do you by any chance have a suggestion that I'd need to follow? Thank you!

**Reply by erez-tom** [reply]
These errors are usually indicative of the deeper issue of failing to load the MuJoCo library.  Do you see the binary in the package directory?  Can you confirm that it is loading successfully?
